### PR TITLE
nrf5340_audio: enable passive scanning to find name in scan response

### PR DESCRIPTION
**Pull request by mringwal**
I'm trying to connect to OnePlus Buds Pro 2 and finally figured out why the unicast demo didn't connect to it: the device name is not in the regular advertising, instead it's in the scan response . This PR enables passive scanning.

### CODE REVIEW

**Comment by alexsven** [comment]
Hey @mringwal, thank you for the PR. Looking at the changes it doesn't really match the commit msg; you are enabling active scan, not passive. We can't hard code that into the project, but it's a good point that we should be able to configure different scan types. 

I suggest adding a Kconfig for this.

**Review by alexsven — CHANGES_REQUESTED** [review]

**Comment by mringwal** [comment]
HI @alexsven Sorry for the line re-ordering. Fixed in the latest force-push. 
I've also updated the PR title to reflect the changes.

If active scan is enabled, it will find the device name in both advertising data as well as scan data. 

Do you see a problem in using active scan in general?
If yes, a Kconfig flag could be used as suggested by you.

**Comment by alexsven** [comment]
> HI @alexsven Sorry for the line re-ordering. Fixed in the latest force-push. I've also updated the PR title to reflect the changes.
> 
> If active scan is enabled, it will find the device name in both advertising data as well as scan data.
> 
> Do you see a problem in using active scan in general? If yes, a Kconfig flag could be used as suggested by you.

It's using a lot more power than the passive  scan. So if the active scan is not needed we should use passive scanning

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c [hunk 1]
```diff
@@ -1009,8 +1009,8 @@ static void on_device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
 			bt_foreach_bond(BT_ID_DEFAULT, bond_connect, (void *)addr);
 		}
 		return;
-	} else if ((type == BT_GAP_ADV_TYPE_ADV_IND || type == BT_GAP_ADV_TYPE_EXT_ADV) &&
-		   (bonded_num < CONFIG_BT_MAX_PAIRED)) {
+	} else if ((type == BT_GAP_ADV_TYPE_ADV_IND || type == BT_GAP_ADV_TYPE_EXT_ADV || (type == BT_GAP_ADV_TYPE_SCAN_RSP)) &&
+	   		(bonded_num < CONFIG_BT_MAX_PAIRED)) {
```
Comment (alexsven) [inline applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:1013]: Either add parenthesis to all of the 'type == BT_GAP...' or remove the ones for your addition, so that it looks more consistent. There is also the formatting, we use clang for out files, and the line shouldn't be more than 100 characters. I've posted the correct formatting below
```suggestion
	} else if ((type == BT_GAP_ADV_TYPE_ADV_IND || type == BT_GAP_ADV_TYPE_EXT_ADV ||
		    type == BT_GAP_ADV_TYPE_SCAN_RSP) &&
		   (bonded_num < CONFIG_BT_MAX_PAIRED)) {
```

**Comment by mringwal** [comment]
Ok. As zephyr / west / Kconfig beginner, I've added a Kconfig option to enable active scan and then used it to select active scan if requested.

**Comment by alexsven** [comment]
> Ok. As zephyr / west / Kconfig beginner, I've added a Kconfig option to enable active scan and then used it to select active scan if requested.

Very nice! Two final comments is that the Kconfig choice you added should be in the bluetooth Kconfig. So in nrf/applications/nrf5340_audio/src/bluetooth/Kconfig. Also, the new define from le_audio.h should be applied to le_audio_bis_headset.c as well. 

Other than that this looks great, I like the switch you added to the adv_parse.

**Comment by mringwal** [comment]
Thanks for the review. I've moved the kconfig flag to bluetooth. For broadcast, different flags/macros are used so I had to introduce another one. 

Btw. the broadcast gateway doesn't build for me as `CONFIG_BT_AUDIO_BROADCAST_SRC_SUBGROUP_COUNT` is not defined, which doesn't seem to be related to my changes that enable active scanning.

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c [hunk 2]
```diff
@@ -1003,16 +1003,23 @@ static void ad_parse(struct net_buf_simple *p_ad, const bt_addr_le_t *addr)
 static void on_device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
 			    struct net_buf_simple *p_ad)
 {
-	/* Direct advertising has no payload, so no need to parse */
-	if (type == BT_GAP_ADV_TYPE_ADV_DIRECT_IND) {
+	switch(type){
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:1006]: Nice change. This is easier to read.

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c [hunk 3]
```diff
@@ -1003,16 +1003,23 @@ static void ad_parse(struct net_buf_simple *p_ad, const bt_addr_le_t *addr)
...
+	case BT_GAP_ADV_TYPE_ADV_DIRECT_IND:
+		/* Direct advertising has no payload, so no need to parse */
 		if (bonded_num) {
 			bt_foreach_bond(BT_ID_DEFAULT, bond_connect, (void *)addr);
 		}
-		return;
-	} else if ((type == BT_GAP_ADV_TYPE_ADV_IND || type == BT_GAP_ADV_TYPE_EXT_ADV) &&
-		   (bonded_num < CONFIG_BT_MAX_PAIRED)) {
+		break;
+	case BT_GAP_ADV_TYPE_ADV_IND:
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:1013]: The code standard requires that /* Fall through */ is added to these cases to underline that this is intentional.

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c [hunk 4]
```diff
@@ -1003,16 +1003,23 @@ static void ad_parse(struct net_buf_simple *p_ad, const bt_addr_le_t *addr)
...
+	case BT_GAP_ADV_TYPE_EXT_ADV:
+	case BT_GAP_ADV_TYPE_SCAN_RSP:
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:1015]: This SCAN_RSP type did not previously end up in this if. Can you comment?

**Review by koffes — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/bluetooth/le_audio.h [hunk 5]
```diff
@@ -16,6 +16,14 @@
 
 #define LE_AUDIO_SDU_SIZE_OCTETS(bitrate) (bitrate / (1000000 / CONFIG_AUDIO_FRAME_DURATION_US) / 8)
 
+#if (CONFIG_SCAN_MODE_ACTIVE)
+#define BT_LE_SCAN_TYPE    BT_LE_SCAN_TYPE_ACTIVE
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/le_audio.h:20]: Since these defines are set here, we don't recommend using BT_LE_* as these are defined elsewhere.

**Review by koffes — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/bluetooth/le_audio.h [hunk 6]
```diff
@@ -16,6 +16,14 @@
...
+#define BT_LE_SCAN_PARAMS  BT_LE_SCAN_ACTIVE
+#else
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/le_audio.h:22]: Some nitpick, but these changes make the code more resilient from outside/upstream changes.
```suggestion
#elif (CONFIG_SCAN_MODE_PASSIVE)
```
And then an #error .

**Review by koffes — CHANGES_REQUESTED** [review]

**Comment by alexsven** [comment]
> Thanks for the review. I've moved the kconfig flag to bluetooth. For broadcast, different flags/macros are used so I had to introduce another one.
> 
> Btw. the broadcast gateway doesn't build for me as `CONFIG_BT_AUDIO_BROADCAST_SRC_SUBGROUP_COUNT` is not defined, which doesn't seem to be related to my changes that enable active scanning.

That might be a case of mismatched versions of zephyr. Using 'west update', as long as you haven't changed west.yml should fix that. I know that they have renamed a lot of the CONFIGs in zephyr upstream lately.

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c (hunk 2 above)
Comment (mringwal) [reply applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:1006]: Thanks

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c (hunk 4 above)
Comment (mringwal) [reply applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:1015]: I'll move the SCAN_RSP addition to a later commit then.

File: applications/nrf5340_audio/src/bluetooth/le_audio.h (hunk 5 above)
Comment (mringwal) [reply applications/nrf5340_audio/src/bluetooth/le_audio.h:20]: OK, I'll use a project specific prefix then.

File: applications/nrf5340_audio/src/bluetooth/le_audio.h (hunk 6 above)
Comment (mringwal) [reply applications/nrf5340_audio/src/bluetooth/le_audio.h:22]: Can add that, too.

**Comment by mringwal** [comment]
I hope I caught all requested changes.

File: applications/nrf5340_audio/src/audio/streamctrl.c [hunk 7]
```diff
@@ -365,6 +365,7 @@ static void le_audio_evt_handler(enum le_audio_evt_type event)
 	int ret;
 
 	LOG_DBG("Received event = %d, current state = %d", event, strm_state);
+	uint32_t pres_delay;
```
Comment (rick1082) [inline applications/nrf5340_audio/src/audio/streamctrl.c:368]: Would be great to move line 368 to line 366 and leave a blank after that.

**Review by rick1082 — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/audio/streamctrl.c (hunk 7 above)
Comment (mringwal) [reply applications/nrf5340_audio/src/audio/streamctrl.c:368]: Hi @rick1082 - I've made no changes to that. I only wanted to add an option to enable active scanning. 

Is it easier for you if I make this change in this PR now?

Comment (mringwal) [reply applications/nrf5340_audio/src/audio/streamctrl.c]: Done. I forgot that I made the changes to get it to compile. Yes, looks better that way.

File: applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c [hunk 8]
```diff
@@ -633,7 +633,7 @@ int le_audio_enable(le_audio_receive_cb recv_cb, le_audio_timestamp_cb timestmp_
 
 	LOG_INF("Scanning for broadcast sources");
 
-	ret = bt_audio_broadcast_sink_scan_start(BT_LE_SCAN_PASSIVE);
+	ret = bt_audio_broadcast_sink_scan_start(BT_LE_SCAN_PARAMS);
```
Comment (alexsven) [inline applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c:636]:
```suggestion
	ret = bt_audio_broadcast_sink_scan_start(NRF5340_AUDIO_GATEWAY_SCAN_PARAMS);
```

Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c:636]: It was failing compile because of this

**Review by alexsven — CHANGES_REQUESTED** [review]

**Comment by alexsven** [comment]
Commits should be squashed to a single commit and to pass compliance it needs to follow the guidelines outlined here: https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines

You can also click the Compliance Details to see the things it is complaining about. 

TLDR: Squash into one commit, header should be "applications: nrf5340_audio: Add Kconfig for scan type". There are also some whitespaces at end of line in some of the files.

File: applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c (hunk 8 above)
Comment (mringwal) [reply applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c]: Thanks. Missed that one. Fixed in the last force-push/squash action. Let's see if there's still compliance issues.

**Comment by alexsven** [comment]
> The Compliance test doesn't start because of "Waiting for status to be reported". No idea, clicking 're-request review'. Other options?

I think it's because I have to approve that the CI will run, and it has to be done every time there is a new commit. Approved that it will be run now.

**Review by rick1082 — APPROVED** [review]

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c (hunk 4 above)
Comment (koffes) [reply applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:1044]: The SCAN_RSP

**Review by koffes — APPROVED** [review]

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c (hunk 4 above)
Comment (mringwal) [reply applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:1044]: The SCAN_RSP was actually done in a later commit. After the squash, this detail was gone. :)

**Comment by alexsven** [comment]
@mringwal: I've asked to get it merged now, thank you for your contribution.
